api/src/main/java/org/apache/iceberg/io/CloseableIterable.java

```diff
@@ -84,6 +113,20 @@ protected boolean shouldKeep(E item) {
     }, iterable);
   }
 
+  static <E> CloseableIterable<E> filter(MetricsContext.Counter<?> counter, CloseableIterable<E> iterable,
+      Predicate<E> pred) {
```

Comment:
Should there be a version that increments a `skip` counter?

I'd also prefer to add the counter support directly in `FilterIterator` rather than using a predicate. There is no guarantee that the predicate is called only once per value. Putting the increment logic inside `FilterIterator#advance` is better.

Reply:
If you want, we can also make a subclass that has callbacks for incrementing, so that we don't need to use NOOP counters when counters aren't supplied.

Reply:
this version is actually incrementing a `skip` counter. I just forgot adding some javadocs to it.

Based on the current contract of [FilterIterator](https://github.com/apache/iceberg/blob/bbcc2fbf6db6f5386fa9e6d169a4cc19d7883ef2/api/src/main/java/org/apache/iceberg/io/FilterIterator.java#L67), it seems that `shouldKeep` is only called once. I tried to keep the changes for counting skipped elements to a minimum, so is this somethind we'd want to do in a follow-up PR?